Source/Diagnostics/SliceDiagnostic.cpp

```diff
@@ -0,0 +1,490 @@
+#include "SliceDiagnostic.H"
+#include <AMReX_MultiFabUtil.H>
+#include <AMReX_PlotFileUtil.H>
+#include <AMReX_FillPatchUtil_F.H>
+
+#include <WarpX.H>
+
+
+
+
+/* \brief
+ *  The functions creates the slice for diagnostics based on the user-input.
+ *  The slice can be 1D, 2D, or 3D and it inherts the index type of the underlying data. 
+ *  The implementation assumes that the slice is aligned with the coordinate axes. 
+ *  The input parameters are modified if the user-input does not comply with requirements of coarsenability or if the slice extent is not contained within the simulation domain. 
+ *  First a slice multifab (smf) with cell size equal to that of the simulation grid is created such that it extends from slice.dim_lo to slice.dim_hi and shares the same index space as the source multifab (mf) 
+ *  The values are copied from src mf to dst smf using amrex::ParallelCopy
+ *  If interpolation is required, then on the smf, using data points stored in the ghost cells, the data in interpolated. 
+ *  If coarsening is required, then a coarse slice multifab is generated (cs_mf) and the 
+ *  values of the refined slice (smf) is averaged down to obtain the coarse slice. 
+ *  \param slice_realbox defines the extent of the slice 
+ *  \param slice_cr_ratio provides the coarsening ratio for diagnostics 
+ *  \param SliceType is the index type of the slice. By default is slice is cell-centered. 
+ *  If the source multifab (mf) has a different index type then SliceType is converted.
+ */
+
+using namespace amrex;
+std::unique_ptr<MultiFab> CreateSlice( const amrex::MultiFab& mf, 
+                          const amrex::Vector<Geometry> &dom_geom, 
+                          amrex::RealBox &slice_realbox, 
+                          amrex::IntVect &slice_cr_ratio )
+{
+    std::unique_ptr<MultiFab> smf;
+    std::unique_ptr<MultiFab> cs_mf;
+
+    Vector<int> slice_ncells(AMREX_SPACEDIM);
+    int nghost = 1;
+    int nlevels = dom_geom.size();
+    int ncomp = (mf).nComp();
+    
+    IntVect cr_ratio(AMREX_D_DECL(0,0,0));
+    const auto conversionType = (mf).ixType();
+    IntVect SliceType(AMREX_D_DECL(0,0,0));
+ 
+    for (int idim = 0; idim < AMREX_SPACEDIM; ++idim )
+    {
+        if ( conversionType.nodeCentered(idim))
+        {
+           SliceType[idim] = 1;
+        }
+    }
+
+    const RealBox& real_box = dom_geom[0].ProbDomain();
+    RealBox slice_cc_nd_box;
+    int slice_grid_size = 32;
+         
+    bool interpolate = false;
+    bool coarsen = false;
+
+    // same index space as domain //
+    IntVect slice_lo(AMREX_D_DECL(0,0,0));
+    IntVect slice_hi(AMREX_D_DECL(1,1,1));
+    IntVect slice_lo2(AMREX_D_DECL(0,0,0));
```

Comment:
I fine the index of the lower co-ordinate and store it in slice_lo, for which I use the floor. 
But due to round-off errors, since I cannot use "==", I perform a calculation with floor and ceil and if they result in the same integer value, then the point is aligned with the underlying data point, else it requires interpolation. Note that, for cell-centered data, even if the floor and ceil value are the same, if the slice.lo < the first cell-centered data-point, interpolation would still be necessary. Therefore, an additional condition is imposed to flag interp_lo = 1.